src/protocol/boolean/solved_bits.rs

```diff
@@ -0,0 +1,265 @@
+    IsPLessThanB,
+    RevealC,
+}
+
+impl crate::protocol::Substep for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::RandomFields => "random_fields",
+            Self::ConvertShares => "convert_shares",
+            Self::IsPLessThanB => "is_p_less_than_b",
+            Self::RevealC => "reveal_c",
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::SolvedBits;
+    use crate::{
+        error::Error,
+        ff::{Field, Fp31, Fp32BitPrime},
+        protocol::{context::ProtocolContext, QueryId, RecordId},
+        secret_sharing::Replicated,
+        test_fixture::{
+            bits_to_value, make_contexts, make_world, validate_and_reconstruct, TestWorld,
+        },
+    };
+    use futures::future::try_join3;
+    use rand::{distributions::Standard, prelude::Distribution};
+
+    async fn random_bits<F: Field>(
+        ctx: [ProtocolContext<'_, Replicated<F>, F>; 3],
+        record_id: RecordId,
+    ) -> Result<Option<(Vec<F>, F)>, Error>
+    where
+        Standard: Distribution<F>,
+    {
+        let [c0, c1, c2] = ctx;
+
+        // Execute
+        let (result0, result1, result2) = try_join3(
+            SolvedBits::execute(c0.bind(record_id), record_id),
+            SolvedBits::execute(c1.bind(record_id), record_id),
+            SolvedBits::execute(c2.bind(record_id), record_id),
+        )
+        .await?;
+
+        // if one of `SolvedBits` calls aborts, then all must have aborted, too
+        if result0.is_none() || result1.is_none() || result2.is_none() {
+            assert!(result0.is_none());
+            assert!(result1.is_none());
+            assert!(result2.is_none());
+            return Ok(None);
+        }
+
+        let (s0, s1, s2) = (result0.unwrap(), result1.unwrap(), result2.unwrap());
+
+        // [b]_B must be the same bit lengths
+        assert_eq!(s0.b_b.len(), s1.b_b.len());
+        assert_eq!(s1.b_b.len(), s2.b_b.len());
+
+        // Reconstruct b_B from ([b_1]_p,...,[b_l]_p) bitwise sharings in F_p
+        let b_b = (0..s0.b_b.len())
+            .map(|i| {
+                let bit = validate_and_reconstruct(&s0.b_b[i], &s1.b_b[i], &s2.b_b[i]);
+                assert!(bit == F::ZERO || bit == F::ONE);
+                bit
+            })
+            .collect::<Vec<_>>();
+
+        // Reconstruct b_P
+        let b_p = validate_and_reconstruct(&s0.b_p, &s1.b_p, &s2.b_p);
+
+        Ok(Some((b_b, b_p)))
+    }
+
+    #[tokio::test]
+    pub async fn fp31() -> Result<(), Error> {
+        let world: TestWorld = make_world(QueryId);
+        let ctx = make_contexts::<Fp31>(&world);
+        let [c0, c1, c2] = ctx;
+
+        // 0.012 secs per invocation. 100 tries = 1.2 secs
+        for i in 0..100 {
+            let record_id = RecordId::from(i);
+            // The chance of this protocol aborting 100 out of 100 tries in Fp31 is < 0.0001%.
+            // Should we test and see if it succeeds at least once?
```

Comment:
Yes, we probably should. Otherwise, imagine a bug that causes this to just return None 100% of the time. This test will still pass.

Reply:
Are you sure that this calculation is correct?  $31^{-100}$ is much less than 1 in a million.  That is, assuming the abort happens 1/31.

Also, one in a million is not really good enough odds for the test.  You should aim for about $2^{-100}$ odds of failure, which is only 21 runs for `Fp31`.  That's faster than this as well.

Reply:
We generate 8 random bits, each bit having a 50/50 chance of being 0 or 1. So, the chance of the number being in the range of Fp31 is, the 3 higher bits need to be 0's and others could be either 0 or 1, which is 1/2^3? (ignoring "30" where the last bit needs to be 0 if other 4 bits are 1's)

So the abort happens about 1 - 1/8 = 88%, and 100 times that is around 1 in a million, I think. For 2^-100, I need 500+ runs?

Reply:
See above.  You shouldn't need that many runs.  The odds of failure can easily be much lower.